homeassistant/components/switchbot/cover.py

```diff
@@ -0,0 +1,181 @@
+from homeassistant.helpers.restore_state import RestoreEntity
+from homeassistant.helpers.update_coordinator import CoordinatorEntity
+
+from .const import (
+    ATTR_CURTAIN,
+    CONF_RETRY_COUNT,
+    DATA_COORDINATOR,
+    DOMAIN,
+    MANUFACTURER,
+)
+from .coordinator import SwitchbotDataUpdateCoordinator
+
+# Initialize the logger
+_LOGGER = logging.getLogger(__name__)
+PARALLEL_UPDATES = 1
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Set up Switchbot curtain based on a config entry."""
+    coordinator: SwitchbotDataUpdateCoordinator = hass.data[DOMAIN][entry.entry_id][
+        DATA_COORDINATOR
+    ]
+
+    if entry.data[CONF_SENSOR_TYPE] != ATTR_CURTAIN:
+        return
+
+    async_add_entities(
+        [
+            SwitchBotCurtain(
+                coordinator,
+                entry.unique_id,
+                entry.data[CONF_MAC],
+                entry.data[CONF_NAME],
+                entry.data.get(CONF_PASSWORD, None),
+                entry.options[CONF_RETRY_COUNT],
+            )
+        ]
+    )
+
+
+class SwitchBotCurtain(CoordinatorEntity, CoverEntity, RestoreEntity):
+    """Representation of a Switchbot."""
+
+    coordinator: SwitchbotDataUpdateCoordinator
+    _attr_device_class = DEVICE_CLASS_CURTAIN
+    _attr_supported_features = (
+        SUPPORT_OPEN | SUPPORT_CLOSE | SUPPORT_STOP | SUPPORT_SET_POSITION
+    )
+    _attr_assumed_state = True
+
+    def __init__(
+        self,
+        coordinator: SwitchbotDataUpdateCoordinator,
+        idx: str | None,
+        mac: str,
+        name: str,
+        password: str,
+        retry_count: int,
+    ) -> None:
+        """Initialize the Switchbot."""
+        super().__init__(coordinator)
+        self._last_run_success: bool | None = None
+        self._idx = idx
+        self._attr_unique_id = idx
+        self._attr_name = name
+        self._mac = mac
+        self._attr_device_info: DeviceInfo = {
+            "connections": {(dr.CONNECTION_NETWORK_MAC, self._mac)},
+            "name": self._attr_name,
+            "model": self.coordinator.data[self._idx]["modelName"],
+            "manufacturer": MANUFACTURER,
+        }
+        self._device = self.coordinator.switchbot_api.SwitchbotCurtain(
+            mac=mac, password=password, retry_count=retry_count
+        )
+
+    async def async_added_to_hass(self) -> None:
+        """Run when entity about to be added."""
+        await super().async_added_to_hass()
+        last_state = await self.async_get_last_state()
+        if not last_state:
+            return
+        if ATTR_CURRENT_POSITION in last_state.attributes:
+            self._attr_current_cover_position = last_state.attributes[
+                ATTR_CURRENT_POSITION
+            ]
+            self._last_run_success = last_state.attributes["last_run_success"]
+
+    @property
+    def device_state_attributes(self) -> dict:
```

Comment:
```suggestion
    def extra_state_attributes(self) -> dict:
```

Reply:
Also I missed the one in `switch.py` (its should be `extra_state_attributes`) as well.

Reply:
Since `{"last_run_success": self._last_run_success, "mac_address": self._mac}` are both shared between this entity and `sensor.py`, make a base class 'SwitchBotEntity` and do something like

`return {**super().extra_state_attributes, "switch_mode": self.coordinator.data[self._idx]["data"]["switchMode"]}` in `switch.py`

Reply:
Examples

https://github.com/home-assistant/core/blob/dev/homeassistant/components/bond/entity.py#L25
https://github.com/home-assistant/core/blob/dev/homeassistant/components/doorbird/entity.py#L25

Reply:
Thank you for the examples. I have added an entity.py file and removed the duplicate code.